File: models/TranslationRotationModel.m
classdef TranslationRotationModel < BaseModel
    %
    % State vector definition
    %   1-3 : position
    %   4-6 : velocity
    %   7-10 : orientation
    %   11-13 : accel bias
    %   14-16 : gyro bias
    %   17 : baro bias
    %
    % Control vector definition
    %   1-3 : acceleration
    %   4-6 : angular velocity

    % Noise vector definition
    %   1-3 : acceleration noise
    %   4-6 : gyro noise
    %   7 : barometer noise
    %   8-10 : acceleration bias noise
    %   11-13 : gyro bias noise
    %   14 : barometer bias noise

    properties (Constant)
        additive_noise = 1e-6;

        % Time has been taken into account within the biases/noises/drifts
        % units = m/s
        accel_bias = [0.0029394893123127377, -0.0009818539510592773, 0.0028762637247315066];
        accel_noise = 4.7358607479753485e-09;
        accel_drift = 3.314312818032142e-10;

        %units = rad/s
        gyro_bias = [0.0038284331173227743, -0.001784190927555866, -0.002920534852288187];
        gyro_noise = 2.5617707075843632e-08;
        gyro_drift = 3.9979150848914756e-10;

        % units = m
        baro_bias = 399.23657624056926;
        baro_noise = 0.014769875002697693;
        baro_drift = 6.282361435771973e-05;

    end

    methods
        function [x_init, P_init] = get_init_state(obj)
            x_init = [
                0
                0
                0

                0
                0
                0

                1
                0
                0
                0

                obj.accel_bias(1)
                obj.accel_bias(2)
                obj.accel_bias(3)

                obj.gyro_bias(1)
                obj.gyro_bias(2)
                obj.gyro_bias(3)

                obj.baro_bias];

            P_init = diag(1e-9 * ones(1, 17));
        end

        function x_new = compute_x_new(obj, x, u)
            dt = obj.dt;
            delta_acc = Utils.body_to_inertial_frame(x(7:10)', (dt*u(1:3)' - x(11:13)));
            delta_q = quaternion([1,  (dt * u(4:6) - x(14:16)') / 2]);
            dx = [
                x(4) * dt
                x(5) * dt
                x(6) * dt

                delta_acc(1)
                delta_acc(2)
                delta_acc(3)

                0 % delta_q must be computed separately
                0
                0
                0

                0
                0
                0

                0
                0
                0

                0];

            x_new = x + dx;
            x_new(7:10) = compact(normalize(quaternion(x(7:10)') * delta_q));
        end

        function F = get_F_matrix(obj, x, u)

            F = zeros(17, 17);

            q0 = x(7); q1 = x(8); q2 = x(9); q3 = x(10);
            dvx = u(1) * obj.dt; dvy = u(2) * obj.dt; dvz = u(3) * obj.dt;
            dax = u(4) * obj.dt; day = u(5) * obj.dt; daz = u(6) * obj.dt;
            dvx_b = x(11); dvy_b = x(12); dvz_b = x(13);
            dax_b = x(14); day_b = x(15); daz_b = x(16);

            % dp^dot
            F(1:3, 4:6) = eye(3) * obj.dt;

            %dv^dot
            F(4, :) = [0, 0, 0, 0, 0, 0, 2*q0*(dvx - dvx_b) - 2*q3*(dvy - dvy_b) + 2*q2*(dvz - dvz_b), 2*q1*(dvx - dvx_b) + 2*q2*(dvy - dvy_b) + 2*q3*(dvz - dvz_b), 2*q1*(dvy - dvy_b) - 2*q2*(dvx - dvx_b) + 2*q0*(dvz - dvz_b), 2*q1*(dvz - dvz_b) - 2*q0*(dvy - dvy_b) - 2*q3*(dvx - dvx_b), 0, 0, 0, - q0^2 - q1^2 + q2^2 + q3^2, 2*q0*q3 - 2*q1*q2, - 2*q0*q2 - 2*q1*q3, 0];
            F(5, :) = [0, 0, 0, 0, 0, 0, 2*q3*(dvx - dvx_b) + 2*q0*(dvy - dvy_b) - 2*q1*(dvz - dvz_b), 2*q2*(dvx - dvx_b) - 2*q1*(dvy - dvy_b) - 2*q0*(dvz - dvz_b), 2*q1*(dvx - dvx_b) + 2*q2*(dvy - dvy_b) + 2*q3*(dvz - dvz_b), 2*q0*(dvx - dvx_b) - 2*q3*(dvy - dvy_b) + 2*q2*(dvz - dvz_b), 0, 0, 0, - 2*q0*q3 - 2*q1*q2, - q0^2 + q1^2 - q2^2 + q3^2, 2*q0*q1 - 2*q2*q3,  0];
            F(6, :) = [0, 0, 0, 0, 0, 0, 2*q1*(dvy - dvy_b) - 2*q2*(dvx - dvx_b) + 2*q0*(dvz - dvz_b), 2*q3*(dvx - dvx_b) + 2*q0*(dvy - dvy_b) - 2*q1*(dvz - dvz_b), 2*q3*(dvy - dvy_b) - 2*q0*(dvx - dvx_b) - 2*q2*(dvz - dvz_b), 2*q1*(dvx - dvx_b) + 2*q2*(dvy - dvy_b) + 2*q3*(dvz - dvz_b), 0, 0, 0,   2*q0*q2 - 2*q1*q3, - 2*q0*q1 - 2*q2*q3, - q0^2 + q1^2 + q2^2 - q3^2, 0];

            q=quaternion(x(7:10)');
            q=normalize(q);
            F(4:6, 14:16) = -quat2rotm(q);

            %dq^dot
            F(7, :) = [0, 0, 0, 0, 0, 0, 0,dax_b/2 - dax/2,day_b/2 - day/2, daz_b/2 - daz/2, 0, 0, 0, q1/2, q2/2, q3/2, 0];
            F(8, :) = [0, 0, 0, 0, 0, 0, dax/2 - dax_b/2, 0,daz/2 - daz_b/2, day_b/2 - day/2, 0, 0, 0, -q0/2, q3/2, -q2/2, 0];
            F(9, :) = [0, 0, 0, 0, 0, 0, day/2-day_b/2, daz_b/2-daz/2, 0, dax/2-dax_b/2, 0, 0, 0, -q3/2, -q0/2, q1/2, 0];
            F(10, :) = [0, 0, 0, 0, 0, 0, daz/2-daz_b/2,day/2-day_b/2,dax_b/2-dax/2,0, 0, 0, 0, q2/2, -q1/2, -q0/2, 0];
        end

        function Q = get_Q_matrix(obj, x, u, w)
            dvxCov = w(1);
            dvyCov = w(2);
            dvzCov = w(3);
            daxCov = w(4);
            dayCov = w(5);
            dazCov = w(6);

            q0 = x(7);
            q1 = x(8);
            q2 = x(9);
            q3 = x(10);

            Q = zeros(17, 17);

            Q(4, :) = [0, 0, 0, dvyCov*(2*q0*q3 - 2*q1*q2)^2 + dvzCov*(2*q0*q2 + 2*q1*q3)^2 + dvxCov*(q0^2 + q1^2 - q2^2 - q3^2)^2, dvxCov*(2*q0*q3 + 2*q1*q2)*(q0^2 + q1^2 - q2^2 - q3^2) - dvyCov*(2*q0*q3 - 2*q1*q2)*(q0^2 - q1^2 + q2^2 - q3^2) - dvzCov*(2*q0*q1 - 2*q2*q3)*(2*q0*q2 + 2*q1*q3), dvzCov*(2*q0*q2 + 2*q1*q3)*(q0^2 - q1^2 - q2^2 + q3^2) - dvxCov*(2*q0*q2 - 2*q1*q3)*(q0^2 + q1^2 - q2^2 - q3^2) - dvyCov*(2*q0*q1 + 2*q2*q3)*(2*q0*q3 - 2*q1*q2), 0, 0, 0, 0,  0, 0, 0,  0, 0, 0, 0];
            Q(5, :) = [0, 0, 0, dvxCov*(2*q0*q3 + 2*q1*q2)*(q0^2 + q1^2 - q2^2 - q3^2) - dvyCov*(2*q0*q3 - 2*q1*q2)*(q0^2 - q1^2 + q2^2 - q3^2) - dvzCov*(2*q0*q1 - 2*q2*q3)*(2*q0*q2 + 2*q1*q3), dvxCov*(2*q0*q3 + 2*q1*q2)^2 + dvzCov*(2*q0*q1 - 2*q2*q3)^2 + dvyCov*(q0^2 - q1^2 + q2^2 - q3^2)^2, dvyCov*(2*q0*q1 + 2*q2*q3)*(q0^2 - q1^2 + q2^2 - q3^2) - dvzCov*(2*q0*q1 - 2*q2*q3)*(q0^2 - q1^2 - q2^2 + q3^2) - dvxCov*(2*q0*q2 - 2*q1*q3)*(2*q0*q3 + 2*q1*q2), 0, 0, 0, 0,  0, 0, 0,  0, 0, 0, 0];
            Q(6, :) = [0, 0, 0, dvzCov*(2*q0*q2 + 2*q1*q3)*(q0^2 - q1^2 - q2^2 + q3^2) - dvxCov*(2*q0*q2 - 2*q1*q3)*(q0^2 + q1^2 - q2^2 - q3^2) - dvyCov*(2*q0*q1 + 2*q2*q3)*(2*q0*q3 - 2*q1*q2), dvyCov*(2*q0*q1 + 2*q2*q3)*(q0^2 - q1^2 + q2^2 - q3^2) - dvzCov*(2*q0*q1 - 2*q2*q3)*(q0^2 - q1^2 - q2^2 + q3^2) - dvxCov*(2*q0*q2 - 2*q1*q3)*(2*q0*q3 + 2*q1*q2), dvxCov*(2*q0*q2 - 2*q1*q3)^2 + dvyCov*(2*q0*q1 + 2*q2*q3)^2 + dvzCov*(q0^2 - q1^2 - q2^2 + q3^2)^2, 0, 0, 0, 0,  0, 0, 0,  0, 0, 0, 0];

            Q(7, :) = [0, 0, 0,  0, 0, 0, (daxCov*q1^2)/4 + (dayCov*q2^2)/4 + (dazCov*q3^2)/4, (dayCov*q2*q3)/4 - (daxCov*q0*q1)/4 - (dazCov*q2*q3)/4, (dazCov*q1*q3)/4 - (dayCov*q0*q2)/4 - (daxCov*q1*q3)/4, (daxCov*q1*q2)/4 - (dayCov*q1*q2)/4 - (dazCov*q0*q3)/4,  0, 0, 0,   0, 0, 0,  0];
            Q(8, :) = [0, 0, 0,  0, 0, 0, (dayCov*q2*q3)/4 - (daxCov*q0*q1)/4 - (dazCov*q2*q3)/4, (daxCov*q0^2)/4 + (dazCov*q2^2)/4 + (dayCov*q3^2)/4, (daxCov*q0*q3)/4 - (dayCov*q0*q3)/4 - (dazCov*q1*q2)/4, (dazCov*q0*q2)/4 - (dayCov*q1*q3)/4 - (daxCov*q0*q2)/4,  0, 0, 0,   0, 0, 0,  0];
            Q(9, :) = [0, 0, 0,  0, 0, 0, (dazCov*q1*q3)/4 - (dayCov*q0*q2)/4 - (daxCov*q1*q3)/4, (daxCov*q0*q3)/4 - (dayCov*q0*q3)/4 - (dazCov*q1*q2)/4, (dayCov*q0^2)/4 + (dazCov*q1^2)/4 + (daxCov*q3^2)/4, (dayCov*q0*q1)/4 - (daxCov*q2*q3)/4 - (dazCov*q0*q1)/4,  0, 0, 0,   0, 0, 0,  0];
            Q(10, :) = [0, 0, 0, 0, 0, 0, (daxCov*q1*q2)/4 - (dayCov*q1*q2)/4 - (dazCov*q0*q3)/4, (dazCov*q0*q2)/4 - (dayCov*q1*q3)/4 - (daxCov*q0*q2)/4, (dayCov*q0*q1)/4 - (daxCov*q2*q3)/4 - (dazCov*q0*q1)/4,(dazCov*q0^2)/4 + (dayCov*q1^2)/4 + (daxCov*q2^2)/4,   0, 0, 0,   0, 0, 0,  0];

        end

        function z_hat  = get_measurement_estimate(obj, x)
            % If using multiple sensors here, h just gets more columns
            z_hat = [x(3) + x(17)];
        end

        function H = get_H_matrix(obj)
            % If using multiple sensors here, H just gets more rows
            H = [...
                0, 0, 1,    0, 0, 0,    0, 0, 0, 0,    0, 0, 0    0, 0, 0,   0];
        end

        function [Qs, w] = generate_noise(obj)
            Fs = 1/obj.dt;
            scale_var = 0.5*(1./(Fs.^2));

            %Qs = diag([obj.additive_noise.*ones(1,10), scale_var.*obj.accel_drift.*ones(1,3), scale_var.* obj.gyro_drift.*ones(1,3), scale_var.* obj.baro_drift.*ones(1,1)]);
            Qs = diag([obj.additive_noise.*ones(1,10), zeros(1, 7)]);

            w = scale_var.*[obj.accel_noise.*ones(1,3), obj.gyro_noise.*ones(1,3), obj.baro_noise.*ones(1,1)];
        end

        function R = get_R_matrix(obj)
            R = 0.1; %obj.baro_noise;
        end

        function x = callback(obj, engine)
            %anti-drift
            x = engine.x;
            x(14:16) = obj.gyro_bias;
        end
    end
end

File: shai/src/models/TranslationRotationModel.cpp
//
// Created by Samuel on 15/05/2023.
//

#include <cmath>
#include <eigen3/Eigen/Geometry>

#include "models/TranslationRotationModel.hpp"

#define NX 17
#define NW 7
#define NZ 1

using namespace shai::models;

Matrix4d hamilton_product_as_matrix(const Vector4d& q){
	Matrix4d mat;
	mat <<  q(0), -q(1), -q(2), -q(3),
			q(1),  q(0), -q(3), q(2),
			q(2), q(3), q(0), -q(1),
			q(3), -q(2), q(1), q(0);

	return mat;
}

Matrix3d rotmat(const Vector4d& q){
	Matrix3d mat;

	double q0 = q(0), q1 = q(1), q2 = q(2), q3 = q(3);

	mat <<  q0*q0+q1*q1-q2*q2-q3*q3, 2*(q1*q2-q0*q3), 2*(q1*q3+q0*q2),
			2*(q1*q2+q0*q3), q0*q0-q1*q1+q2*q2-q3*q3, 2*(q2*q3-q0*q1),
			2*(q1*q3-q0*q2), 2*(q2*q3+q0*q1), q0*q0-q1*q1-q2*q2+q3*q3;
	return mat;
}

Vector4d multquat(const Vector4d& q1, const Vector4d& q2){
	Vector4d q_new;
	q_new(0) = q1(0)*q2(0)    -q1(1)*q2(1)    -q1(2)*q2(2)    -q1(3)*q2(3);
	q_new(0) = q1(0)*q2(1)    +q2(0)*q1(1)    +q1(2)*q2(3)    -q2(2)*q1(3);
	q_new(0) = q1(0)*q2(2)    +q2(0)*q1(2)    -q1(1)*q2(3)    +q2(1)*q1(3);
	q_new(0) = q1(0)*q2(3)    +q2(0)*q1(3)    +q1(1)*q2(2)    -q2(1)*q1(2);
	return q_new;
}

VectorXd TranslationRotationModel::get_init_state() {
	VectorXd x_init = VectorXd::Zero(NX);
	x_init << 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, accel.bias, gyro.bias, baro.bias;
	return x_init;
}

MatrixXd TranslationRotationModel::get_init_cov() {
	MatrixXd P_init = 1e-9 * MatrixXd::Identity(NX, NX);
	return P_init;
}

VectorXd TranslationRotationModel::compute_x_new(const VectorXd &x, const VectorXd &u) {
	Vector3d delta_acc = rotmat(x.segment<4>(6)) * (_dt * u.segment<3>(0) - x.segment<3>(10));
	Vector3d axis = (_dt * u.segment<3>(3) - x.segment<3>(13)) / 2.;

	VectorXd dx(x.size());
	dx << x(3) * _dt,
			x(4) * _dt,
			x(5) * _dt,
			delta_acc(0),
			delta_acc(1),
			delta_acc(2),
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0;

	VectorXd x_new = x + dx;
	Vector4d delta_q{1, axis(0), axis(1), axis(2)};
	Vector4d q = multquat(x.segment<4>(6), delta_q);
	q.normalize();

	x_new.segment<4>(6) = q;

	return x_new;
}

MatrixXd TranslationRotationModel::get_F_matrix(const VectorXd &x, const VectorXd &u) {
	MatrixXd F = MatrixXd::Zero(NX, NX);
	double q0 = x(6), q1 = x(7), q2 = x(8), q3 = x(9);

	double a_x = u(0) * _dt, a_y = u(1) * _dt, a_z = u(2) * _dt;
	double w_x = u(3) * _dt, w_y = u(4) * _dt, w_z = u(5) * _dt;
	double b_acc_x = x(10), b_acc_y = x(11),b_acc_z = x(12);
	double b_gyro_x = x(13), b_gyro_y = x(14), b_gyro_z = x(15);

	F.block<3, 3>(0, 3) << Matrix3d::Identity() * _dt;

	F.row(3) << 0, 0, 0, 0, 0, 0, 2*q0*(a_x - b_acc_x) - 2*q3*(a_y - b_acc_y) + 2*q2*(a_z - b_acc_z), 2*q1*(a_x - b_acc_x) + 2*q2*(a_y - b_acc_y) + 2*q3*(a_z - b_acc_z), 2*q1*(a_y - b_acc_y) - 2*q2*(a_x - b_acc_x) + 2*q0*(a_z - b_acc_z), 2*q1*(a_z - b_acc_z) - 2*q0*(a_y - b_acc_y) - 2*q3*(a_x - b_acc_x), 0, 0, 0, -q0*q0 - q1*q1 + q2*q2 + q3*q3, 2*q0*q3 - 2*q1*q2, -2*q0*q2 - 2*q1*q3, 0;
	F.row(4) << 0, 0, 0, 0, 0, 0, 2*q3*(a_x - b_acc_x) + 2*q0*(a_y - b_acc_y) - 2*q1*(a_z - b_acc_z), 2*q2*(a_x - b_acc_x) - 2*q1*(a_y - b_acc_y) - 2*q0*(a_z - b_acc_z), 2*q1*(a_x - b_acc_x) + 2*q2*(a_y - b_acc_y) + 2*q3*(a_z - b_acc_z), 2*q0*(a_x - b_acc_x) - 2*q3*(a_y - b_acc_y) + 2*q2*(a_z - b_acc_z), 0, 0, 0, -2*q0*q3 - 2*q1*q2, -q0*q0 + q1*q1 - q2*q2 + q3*q3, 2*q0*q1 - 2*q2*q3, 0;
	F.row(5) << 0, 0, 0, 0, 0, 0, 2*q1*(a_y - b_acc_y) - 2*q2*(a_x - b_acc_x) + 2*q0*(a_z - b_acc_z), 2*q3*(a_x - b_acc_x) + 2*q0*(a_y - b_acc_y) - 2*q1*(a_z - b_acc_z), 2*q3*(a_y - b_acc_y) - 2*q0*(a_x - b_acc_x) - 2*q2*(a_z - b_acc_z), 2*q1*(a_x - b_acc_x) + 2*q2*(a_y - b_acc_y) + 2*q3*(a_z - b_acc_z), 0, 0, 0, 2*q0*q2 - 2*q1*q3, -2*q0*q1 - 2*q2*q3, -q0*q0 + q1*q1 + q2*q2 - q3*q3, 0;

	F.row(6) << 0, 0, 0, 0, 0, 0, 0, b_gyro_x/2 - w_x/2, b_gyro_y/2 - w_y/2, b_gyro_z/2 - w_z/2, 0, 0, 0, q1/2, q2/2, q3/2, 0;
	F.row(7) << 0, 0, 0, 0, 0, 0, w_x/2 - b_gyro_x/2, 0, w_z/2 - b_gyro_z/2, b_gyro_y/2 - w_y/2, 0, 0, 0, -q0/2, q3/2, -q2/2, 0;
	F.row(8) << 0, 0, 0, 0, 0, 0, w_y/2 - b_gyro_y/2, b_gyro_z/2 - w_z/2, 0, w_x/2 - b_gyro_x/2, 0, 0, 0, -q3/2, -q0/2, q1/2, 0;
	F.row(9) << 0, 0, 0, 0, 0, 0, w_z/2 - b_gyro_z/2, w_y/2 - b_gyro_y/2, b_gyro_x/2 - w_x/2, 0, 0, 0, 0, q2/2, -q1/2, -q0/2, 0;

	return F;
}

MatrixXd TranslationRotationModel::get_Q_matrix(const VectorXd &x, const VectorXd &u, const VectorXd &w) {
	MatrixXd Q = MatrixXd::Zero(NX, NX);

	double q0 = x(6), q1 = x(7), q2 = x(8), q3 = x(9);

	double dvxCov = w(0), dvyCov = w(1), dvzCov = w(2);
	double daxCov = w(3), dayCov = w(4), dazCov = w(5);

	MatrixXd G = MatrixXd::Zero(NX, NW);
	G.block<3, 3>(3, 0) = rotmat({q0, q1, q2, q3});
	G.block<4, 4>(6, 3) = 0.5*hamilton_product_as_matrix(x.segment<4>(6));
	Q = G * Vector<double, NW>{dvxCov, dvyCov, dvzCov, 0, daxCov, dayCov, dazCov}.asDiagonal() * G.transpose();

	return Q;
}

VectorXd TranslationRotationModel::get_measurement_estimate(const VectorXd &x) {
	VectorXd z = VectorXd::Zero(NZ);
	z << x(2) + x(16);
	return z;
}

MatrixXd TranslationRotationModel::get_H_matrix() {
	MatrixXd H = MatrixXd::Zero(NZ, NX);
	H << 0, 0, 1,  0, 0, 0,  0, 0, 0, 0,  0, 0, 0,  0, 0, 0,  0;
	return H;
}

MatrixXd TranslationRotationModel::get_R_matrix() {
	MatrixXd R = MatrixXd::Zero(NZ, NZ);
	R << baro.noise;
	return R;
}

VectorXd TranslationRotationModel::get_noise_vect() {
	VectorXd w = VectorXd::Zero(NW);
	double Fs = 1./_dt;
	double s =  0.5 * (1. / (Fs * Fs));

	w << Vector3d::Constant(s * accel.noise), Vector3d::Constant(s * gyro.noise), s * baro.noise;
	return w;
}

MatrixXd TranslationRotationModel::get_Qs_matrix() {
	double Fs = 1./_dt;
	double scale_var = 0.5 *(1./ (Fs * Fs));

	Vector3d vel_delta_bias_sigma = Vector3d::Constant(scale_var * accel.drift);
	Vector3d ang_vel_delta_bias_sigma = Vector3d::Constant(scale_var * gyro.drift);
	double pos_delta_bias_sigma = scale_var * baro.drift;

	Eigen::Vector<double, NX> Qs;
	Qs << Vector<double, 10>::Constant(additive_noise), vel_delta_bias_sigma, ang_vel_delta_bias_sigma, pos_delta_bias_sigma;
	return Qs.asDiagonal();
}
